README.md
# @yeisonkirax/ibm-mq-nest

# Table of Contents

- [@yeisonkirax/ibm-mq-nest](#yeisonkiraxibm-mq-nest)
- [Table of Contents](#table-of-contents)
    - [Description](#description)
    - [Requirements](#requirements)
    - [Usage](#usage)
        - [Module Initialization](#module-initialization)
    - [Usage with Interceptors, Guards and Filters](#usage-with-interceptors-guards-and-filters)
    - [Usage with Controllers](#usage-with-controllers)
    - [Receiving Messages](#receiving-messages)
        - [Dealing with the original message](#dealing-with-the-original-message)
    - [Sending Messages](#sending-messages)
        - [Inject the IbmMqConnection](#inject-the-IbmMqConnection)
        - [Publishing Messages](#publishing-Messages)
        - [Waiting for a response](#waiting-for-a-response)

## Description

Inspired
by [@golevelup/nestjs-rabbitmq](https://www.npmjs.com/package/@golevelup/nestjs-rabbitmq)
package, i made this module to integrate ibm queues into our nest
applications. This module use a fork of
[ibmmq](https://www.npmjs.com/package/ibmmq), that permit to subscribe to
multiples queues without block the application.

## Requirements

It is important that you have made the configuration requested by the ibmmq.
You must set up:

* Node version 10.20 or greater. Older versions will no longer work because
  of requirements from the ffi-napi package.
* On platforms other than Windows and Linux x64, you must also install the
  MQ client package.
* The package makes use of the libffi capabilities for direct access to the
  MQ API. For some platforms the libffi and ffi-napi components may need
  additional configuration before they can be build. That package is not
  available at all on z/OS - there is no way that this package can run on that
  platform even with handcrafting build steps.

To see the requirements with more details, please refer
to [ibmmq](https://www.npmjs.com/package/ibmmq)

## Usage

### Module Initialization

You must import and add `IbmMqNestModule` in you appModule and set up the
connection. For now, you can only connect to one queueManager per application.

```typescript
import {IbmMqNestModule} from '@yeisonkirax/ibm-mq-nest';

@Module({
  imports: [IbmMqNestModule.forRoot(IbmMqNestModule, {
    channel: {
      name: "channel name", queueManager: {
        name: "queueManager name",
        host: "the queueManager's url",
        port: "the queueManager's port",
      },
    }, credentials: {
      userId: 'usedId', password: 'password',
    },
  })], controllers: [AppController], providers: [],
})
export class AppModule {
}
```

The parameters that the module receives for its configuration correspond to
the following structure:

```typescript
/**
 * Defining the interface for the credentials' config.
 * @property {string} userId - User's id that can connect to the queueManager.
 * @property {string} password - User's password that can connect to the
 *   queueManager
 */
export interface IMQCredentials {
  /** User's id that can connect to the queueManager **/
  userId: string;

  /** User's password that can connect to the queueManager **/
  password: string;
}

/**
 * Defining the interface for the queueManager.
 * @property {string} name - The name of the queueManager.
 * @property {string} host - The host of the queueManager.
 * @property {string} port - The port of the queueManager.
 */
export interface IMQQueueManager {
  /** The name of the queueManager **/
  name: string;

  /** The host of the queueManager **/
  host: string;

  /** The port of the queueManager **/
  port: string;
}

/**
 * Defining the interface for the channel.
 * @property {string} name - The name of the channel.
 * @property {IMQQueueManager} queueManager - The queueManager's config obj.
 */
export interface IMQChannel {
  /** The name of the channel */
  name: string;
  /** The queueManager's config obj */
  queueManager: IMQQueueManager;
}

/**
 * Defining the interface for the options of the module
 * @property {boolean} [debug] - Enable debug mode. Default is false.
 * @property {number} [maxConsecutiveGets] - How many
 *   messages to get from a queue before trying a different queue. Default
 *   is 100
 * @property {number} [getLoopPollTimeMs] - Milliseconds between each full
 * poll cycle. Default is 10000 (10 seconds).
 * @property {number} [getLoopDelayTimeMs] - Milliseconds to delay after a
 *   partial poll cycle. Default is 250 (1/4 second)
 * @property {MQMatchMessageMode} [matchMessageMode] - Set the message
 * filter when make a get to a queue. You can filter by msg id or none.
 * Default is MSG_ID.
 */
export interface IMQOptions {
  /** Enable debug mode. Default is false. */
  debug?: boolean;

  /** How many messages to get from a queue before trying a different queue.
   * Default is 100
   */
  maxConsecutiveGets?: number;

  /** Milliseconds between each full poll cycle. Default is 10000 (10
   *  seconds). */
  getLoopPollTimeMs?: number;

  /**
   * Milliseconds to delay after a partial poll cycle. Default is 250 (1/4
   * second).
   */
  getLoopDelayTimeMs?: number;
}

/**
 * Defining the interface for the option's module.
 * @property {IMQChannel} channel - channel used to connect to queueManager.
 * @property {IMQCredentials} [credentials] - credentials config object.
 * @property {LoggerService} [logger] - set the logger, default is NestJs Logger
 * @property {IMQOptions} [options] - set the module's options.
 */
export interface IMQConfig {
  /** channel config object */
  channel: IMQChannel;

  /** credentials config object */
  credentials?: IMQCredentials;

  /** set the logger, default is NestJs Logger */
  logger?: LoggerService;

  /** set the module's options. */
  options?: IMQOptions;
}
```

## Usage with Interceptors, Guards and Filters

These functionalities will be implemented in future developments.

## Usage with Controllers

You can use NestJS controllers as handlers.

### WARNING: When using controllers, be aware that no HTTP context is available. You can not combine in a controller rest methods with ibm-mq handlers.

```typescript
import {
  AdditionalInfo, IbmMqConnection, MQObjectType, MQQueueSubscribe,
} from '@yeisonkirax/ibm-mq-nest';

@Controller()
export class AppController {
  constructor() {
  }

  @MQQueueSubscribe({
    targetName: 'DEV.QUEUE.1',
  })
  async queueHandler(msg: string, additionalInfo: AdditionalInfo, error: MQError,) {
  }
}
```

You only need to set the name of the queue that is managed by the
queueManager defined above. You can also subscribe to topics using
`MQTopicSubscribe` decorator. In the same way, you only have to pass the name of
the topic to subscribe.

```typescript
import {
  AdditionalInfo, IbmMqConnection, MQObjectType, MQQueueSubscribe,
} from '@yeisonkirax/ibm-mq-nest';

@Controller()
export class AppController {
  constructor() {
  }

  @MQTopicSubscribe({
    targetName: 'dev/',
  })
  async topicHandler(msg: string, additionalInfo: AdditionalInfo, error: MQError,) {
  }
}
```

## Receiving Messages

A handler can receive:

* `msg`: correspond the message deserialized as `string`.
* `additionalInfo`: correspond the raw data.
* `error`: if exist an error, this parameter will be a `MQError` instance.

### Dealing with the original message

If you need the original message. You can use `additionalInfo` in the
handler. This parameter has the following structure.

```typescript
/**
 * `AdditionalInfo` is an object with a `correlationId` property of type
 * `string`, a `msgId` property of type `string`, a `rawMessage` property of
 * type `Buffer`, and a `messageDescriptor` property of type `MQMD`.
 *
 * The `MQMD` type is defined in the `mq-types` module.
 * @property {string} correlationId - The correlation ID of the message.
 * @property {string} msgId - The message ID of the message that was received.
 * @property {Buffer} rawMessage - The raw message buffer.
 * @property {MQMD} messageDescriptor - This is the MQMD object that is
 *   returned by the MQI call.
 */
export type AdditionalInfo = {
  correlationId: string; msgId: string; rawMessage: Buffer; messageDescriptor: MQMD;
};
```

## Sending Messages

### Inject the IbmMqConnection

All IBM MQ interactions go through the `IbmMqConnection` object. Assuming you
installed and configured the `IbmMqNestModule`, the object can be obtained
through Nest's dependency injection system. Simply require it as a constructor
parameter in a Nest Controller or Service.

```typescript
@Controller()
export class AppController {
  constructor(private readonly ibmMqConnection: IbmMqConnection) {
  }

}
```

### Publishing Messages

If you just want to publish a message onto a queue or topic, use
the `publish` method of the `IbmMqConnection` which has the following signature:

```typescript
/**
 * `PublishMessageOptions` is an object with a `target` property that is either
 * a
 * `TopicOptions` or a `QueueOptions` object, and an optional `replyToQueue`
 * property.
 * @property {TopicOptions | QueueOptions} target - The target topic or queue
 *   to publish the message to.
 * @property {string} replyToQueue - The name of the queue to which the service
 * should send the reply.
 */
export type PublishMessageOptions = {
  /** The target topic or queue to publish the message to */
  target: TopicOptions | QueueOptions;

  /** The name of the queue to which the service should send the reply. */
  replyToQueue?: string;
};

class IbmMqConnection {
  /**
   * It opens a queue or topic, puts a message on it, and closes the queue or
   * topic
   * @param {string} message - The message to be sent.
   * @param {PublishMessageOptions} msgOptions - PublishMessageOptions
   * @returns The return value is a Promise that resolves with the msg id.
   */
  public async publish(message: string, msgOptions: PublishMessageOptions): Promise<{ msgId: Buffer }> {
  }
}
```

For example:

```typescript
this.ibmMqConnection.publish('some-text', {
  target: {
    name: 'DEV.QUEUE.1', type: MQObjectType.QUEUE,
  }, replyToQueue: 'DEV.QUEUE.2'
});
```

### Waiting for a response

If you need wait for a response, you can use the `getMessagesFromQueue` method
of the `IbmMqConnection` which has the following signature:

```typescript
/** Defining a type for the GetMessagesConfig object.
 * @property {QueueOptions} queue - The queue where will get the messages.
 * @property {boolean} filterMessagesById - Enable the filter to get
 * messages by id. If is false, you will not filter the messages.
 * @property {Buffer} [msgId] - The msg id used to filter. If you enable
 * "filterMessagesById", this value must be provided.
 * @property {number} waitInterval - The time in ms that you will wait for
 * the response.
 */
export declare type GetMessagesConfig = {
  /**
   * The queue where will get the messages.
   */
  queue: QueueOptions;

  /**
   * Enable the filter to get messages by id. If is false, you will not
   * filter the messages.
   */
  filterMessagesById: boolean;

  /**
   * The msg id used to filter. If you enable "filterMessagesById", this
   * value must be provided.
   */
  msgId?: Buffer;

  /**
   * The time in ms that you will wait for the messages.
   */
  waitInterval: number;
};

class IbmMqConnection {
  /**
   * Get the messages from a queue.
   * @param {GetMessagesConfig} options - The message to be sent.
   * @returns The return a Promise that resolves with the msg as string or
   * an error.
   */
  public async getMessagesFromQueue(options: GetMessagesConfig,): Promise<string> {
  }
}
```

## IMPORTANT: THIS MODULE IS UNDER DEVELOPMENT SO IS POSSIBLE THAT CAN EXISTS BUGS, PLEASE BE CAREFULLY.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit
guidelines.

# 0.5.0 (2022-11-07)

### Features

- **ibm-mq:** add the initial version, you can subscribe to queues or topics,
  publish messages and configure the connection to queueManager.


src/ibm-mq-nest.module.ts
import { Logger, Module, OnApplicationBootstrap, OnApplicationShutdown } from '@nestjs/common';
import { createConfigurableDynamicRootModule } from '@golevelup/nestjs-modules';
import { IMQConfig, IMQDecoratorOptions, MQObjectType } from './interfaces/mq.interfaces';
import { IBM_MQ_ARGS_METADATA, IBM_MQ_CONFIG_TOKEN, IBM_MQ_HANDLER } from './constants/mq.constants';
import { IbmMqConnectionManager } from './ibm-mq/ibm-mq.connection-manager';
import { IbmMqConnection } from './ibm-mq/ibm-mq.connection';
import { DiscoveryModule, DiscoveryService } from '@golevelup/nestjs-discovery';
import { ExternalContextCreator } from '@nestjs/core/helpers/external-context-creator';
import { groupBy } from 'lodash';
import { IbmMqConnectionFactory, IbmMqParamsFactory } from './factories/mq.factory';
import { ClosePromise } from 'ibmmq';

@Module({
  imports: [DiscoveryModule],
  providers: [],
  exports: [],
})
export class IbmMqNestModule
  extends createConfigurableDynamicRootModule<IbmMqNestModule, IMQConfig>(IBM_MQ_CONFIG_TOKEN, {
    providers: [
      {
        provide: IbmMqConnectionManager,
        useFactory: async (config: IMQConfig, conn: IbmMqConnectionFactory) => {
          const con: IbmMqConnection = await conn.createConnection(config.channel, config.credentials, config.options);
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          this.IbmMqNestModule.connectionManager.addConnection(con);
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          return this.IbmMqNestModule.connectionManager;
        },
        inject: [IBM_MQ_CONFIG_TOKEN, IbmMqConnectionFactory],
      },
      {
        provide: IbmMqConnection,
        useFactory: async (config: IMQConfig): Promise<IbmMqConnection> => {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          return this.IbmMqNestModule.connectionManager.getConnection(
            config.channel.queueManager.name,
          ) as IbmMqConnection;
        },
        inject: [IBM_MQ_CONFIG_TOKEN, IbmMqConnectionManager],
      },
      IbmMqParamsFactory,
      IbmMqConnectionFactory,
    ],
    exports: [IbmMqConnectionManager, IbmMqConnection],
  })
  implements OnApplicationBootstrap, OnApplicationShutdown
{
  private readonly logger = new Logger(IbmMqNestModule.name);
  private static connectionManager = new IbmMqConnectionManager();
  private static bootstrapped = false;

  constructor(
    private readonly discover: DiscoveryService,
    private readonly externalContextCreator: ExternalContextCreator,
    private readonly connectionManager: IbmMqConnectionManager,
    private readonly ibmMqParamsFactory: IbmMqParamsFactory,
  ) {
    super();
  }

  async onApplicationShutdown() {
    this.logger.verbose('Closing IBM MQ Connections');
    const connections = this.connectionManager.getConnections();
    const closeQueuesOpenedPromises: Promise<void>[] = [];
    connections.forEach((connection) => {
      const objectNamesGetMessages = Object.keys(connection.objectsOpenedToGetMessages);
      for (const queueName of objectNamesGetMessages) {
        closeQueuesOpenedPromises.push(ClosePromise(connection.objectsOpenedToGetMessages[queueName], 0));
      }
    });
    await Promise.all(closeQueuesOpenedPromises);

    await Promise.all(this.connectionManager.getConnections().map((connection) => connection.closeConnection()));

    this.connectionManager.clearConnections();
    IbmMqNestModule.bootstrapped = false;
  }

  async onApplicationBootstrap() {
    if (IbmMqNestModule.bootstrapped) {
      return;
    }
    IbmMqNestModule.bootstrapped = true;
    const connections = this.connectionManager.getConnections();
    this.logger.log('Initializing IBM MQ Handlers');
    this.logger.log(
      'Searching for MQ Handlers in Controllers. You can not use NestJS HTTP-Requests in these controllers!',
    );
    let mqMetas = await this.discover.providerMethodsWithMetaAtKey<IMQDecoratorOptions>(IBM_MQ_HANDLER);
    mqMetas = mqMetas.concat(await this.discover.controllerMethodsWithMetaAtKey<IMQDecoratorOptions>(IBM_MQ_HANDLER));
    for (const connection of connections) {
      const metasUsingThisConnection = mqMetas;
      const mqMetasGrouped = groupBy(metasUsingThisConnection, (x) => x.discoveredMethod.parentClass.name);
      const providerKeys = Object.keys(mqMetasGrouped);

      for (const providerKey of providerKeys) {
        this.logger.log(`Registering IBM MQ handlers from ${providerKey}`);
        await Promise.all(
          mqMetasGrouped[providerKey].map(async ({ discoveredMethod, meta }) => {
            const handler = this.externalContextCreator.create(
              discoveredMethod.parentClass.instance,
              discoveredMethod.handler,
              discoveredMethod.methodName,
              IBM_MQ_ARGS_METADATA,
              this.ibmMqParamsFactory,
              undefined, // contextId
              undefined, // inquirerId
              undefined, // options
              'ibm-mq',
            );
            const handlerDisplayName = `${discoveredMethod.parentClass.name} -> ${discoveredMethod.methodName} -> ${meta.targetType} -> ${meta.targetName}}`;
            this.logger.log(handlerDisplayName);
            const handlerName = meta.targetName;
            const handlerObj = {};
            const mergedConfig = {
              ...meta,
              ...handlerObj[handlerName],
            };
            return meta.targetType === MQObjectType.TOPIC
              ? connection.createTopicSubscriber(handler, mergedConfig, discoveredMethod.methodName)
              : connection.createQueueSubscriber(handler, mergedConfig, discoveredMethod.methodName);
          }),
        );
      }
    }
  }
}


src/factories/mq.factory.ts
import { ParamData } from '@nestjs/common';
import { IBM_MQ_HEADER_TYPE, IBM_MQ_PARAM_TYPE, IBM_MQ_REQUEST_TYPE } from '../constants/mq.constants';
import { IMQChannel, IMQCredentials, IMQOptions } from '../interfaces/mq.interfaces';
import { IbmMqConnection } from '../ibm-mq/ibm-mq.connection';
import { isObject } from 'class-validator';

export class IbmMqParamsFactory {
  public exchangeKeyForValue(type: number, data: ParamData, args: any[]) {
    if (!args) {
      return null;
    }

    let index = 0;
    if (type === IBM_MQ_PARAM_TYPE) {
      index = 0;
    } else if (type === IBM_MQ_REQUEST_TYPE) {
      index = 1;
    } else if (type === IBM_MQ_HEADER_TYPE) {
      index = 2;
    }

    return data && !isObject(data) ? args[index]?.[data] : args[index];
  }
}

/* It creates a new instance of the IbmMqConnection class, initializes it, and
 returns it */
export class IbmMqConnectionFactory {
  async createConnection(channel: IMQChannel, credentials: IMQCredentials, options: IMQOptions) {
    const connection = new IbmMqConnection(channel, credentials, options);
    await connection.init();
    return connection;
  }
}


src/decorators/mq.decorators.ts
import { IMQDecoratorOptions, IMQMessageHandler, MQObjectType } from '../interfaces/mq.interfaces';
import { applyDecorators, SetMetadata } from '@nestjs/common';
import { makeInjectableDecorator } from '@golevelup/nestjs-common';
import { IBM_MQ_CONFIG_TOKEN, IBM_MQ_HANDLER } from '../constants/mq.constants';

/**
 * It takes an object of type T, which is a subset of IMQHandlerConfig, and
 * returns a function that takes an object of type Pick<IMQHandlerConfig,
 * Exclude<keyof IMQHandlerConfig, keyof T>>, which is a subset of
 * IMQHandlerConfig that excludes the properties of T
 * @param {T} input - T - This is the input parameter that you will pass
 * to the
 * decorator.
 */
export const makeIBMMQDecorator =
  <T extends Partial<IMQDecoratorOptions>>(input: T) =>
  (config: Pick<IMQDecoratorOptions, Exclude<keyof IMQDecoratorOptions, keyof T>>) =>
    applyDecorators(SetMetadata(IBM_MQ_HANDLER, { ...input, ...config }));

/* A decorator that is used to subscribe to a topic or queue. */
export const MQHandler = (config: IMQMessageHandler) => (target, key, descriptor) =>
  SetMetadata(IBM_MQ_HANDLER, config)(target, key, descriptor);

/* A decorator that is used to subscribe to a topic. */
export const MQTopicSubscribe = makeIBMMQDecorator({
  targetType: MQObjectType.TOPIC,
});
/* A decorator that is used to subscribe to a queue. */
export const MQQueueSubscribe = makeIBMMQDecorator({
  targetType: MQObjectType.QUEUE,
});
export const InjectMQConfig = makeInjectableDecorator(IBM_MQ_CONFIG_TOKEN);


src/interfaces/mq.interfaces.ts
import { LoggerService } from '@nestjs/common';

export enum MQObjectType {
  TOPIC = 'TOPIC',
  QUEUE = 'QUEUE',
}

export enum MatchMessageMode {
  MSG_ID = 'MSG_ID',
  NONE = 'NONE',
}

/**
 * represents the type of the handler.
 * */
export declare type MQHandlerType = MQObjectType.TOPIC | MQObjectType.QUEUE;

/**
 * represents the ways to get messages from a queue or topic
 * */
export declare type MQMatchMessageMode = MatchMessageMode.MSG_ID | MatchMessageMode.NONE;

export declare type MQHandlers = Record<string, any>;

/**
 * Defining the interface for the channel.
 * @property {string} name - The name of the channel.
 * @property {IMQQueueManager} queueManager - The queueManager's config obj.
 */
export interface IMQChannel {
  /** The name of the channel */
  name: string;
  /** The queueManager's config obj */
  queueManager: IMQQueueManager;
}

/**
 * Defining the interface for the queueManager.
 * @property {string} name - The name of the queueManager.
 * @property {string} host - The host of the queueManager.
 * @property {string} port - The port of the queueManager.
 */
export interface IMQQueueManager {
  /** The name of the queueManager **/
  name: string;

  /** The host of the queueManager **/
  host: string;

  /** The port of the queueManager **/
  port: string;
}

/**
 * Defining the interface for the credentials' config.
 * @property {string} userId - User's id that can connect to the queueManager.
 * @property {string} password - User's password that can connect to the
 *   queueManager
 */
export interface IMQCredentials {
  /** User's id that can connect to the queueManager **/
  userId: string;

  /** User's password that can connect to the queueManager **/
  password: string;
}

/**
 * Defining the interface for the options of the module
 * @property {boolean} [debug] - Enable debug mode. Default is false.
 * @property {number} [maxConsecutiveGets] - How many
 *   messages to get from a queue before trying a different queue. Default
 *   is 100
 * @property {number} [getLoopPollTimeMs] - Milliseconds between each full
 * poll cycle. Default is 10000 (10 seconds).
 * @property {number} [getLoopDelayTimeMs] - Milliseconds to delay after a
 *   partial poll cycle. Default is 250 (1/4 second)
 * @property {MQMatchMessageMode} [matchMessageMode] - Set the message
 * filter when make a get to a queue. You can filter by msg id or none.
 * Default is MSG_ID.
 */
export interface IMQOptions {
  /** Enable debug mode. Default is false. */
  debug?: boolean;

  /** How many messages to get from a queue before trying a different queue.
   * Default is 100
   */
  maxConsecutiveGets?: number;

  /** Milliseconds between each full poll cycle. Default is 10000 (10
   *  seconds). */
  getLoopPollTimeMs?: number;

  /**
   * Milliseconds to delay after a partial poll cycle. Default is 250 (1/4
   * second).
   */
  getLoopDelayTimeMs?: number;
}

/**
 * Defining the interface for the option's module.
 * @property {IMQChannel} channel - channel used to connect to queueManager.
 * @property {IMQCredentials} [credentials] - credentials config object.
 * @property {LoggerService} [logger] - set the logger, default is NestJs Logger
 * @property {IMQOptions} [options] - set the module's options.
 */
export interface IMQConfig {
  /** channel config object */
  channel: IMQChannel;

  /** credentials config object */
  credentials?: IMQCredentials;

  /** set the logger, default is NestJs Logger */
  logger?: LoggerService;

  /** set the module's options. */
  options?: IMQOptions;
}

/**
 * Defining the interface for the options of the message handler.
 * @property {string} queueManagerName - Name of the connected queue manager.
 * @property {string} channelName - Name of the channel used.
 * @property {string} targetName - Name of the topic o queue where the
 * message will be published.
 * @property {MQObjectType} targetType -Type of the mqObject (queue or topic)
 *   that will be published the message
 */
export interface IMQMessageHandler {
  /** Name of the connected queue manager */
  queueManagerName: string;

  /** Name of the channel used */
  channelName: string;

  /** Name of the topic o queue where the
   message will be published. */
  targetName: string;

  /** Type of the mqObject (queue or topic)
   that will be published the message */
  targetType: MQObjectType;
}

export interface IMQDecoratorOptions {
  /** Name of the topic o queue where the message will be published.
   */
  targetName: string;

  /**
   * Type of the mqObject (queue or topic) that will be published the message
   */
  targetType: MQObjectType;
}
